src/dfx/src/commands/canister/info.rs

```diff
@@ -23,22 +23,29 @@ pub async fn exec(env: &dyn Environment, opts: InfoOpts) -> DfxResult {
         .ok_or_else(|| anyhow!("Cannot get HTTP client from environment."))?;
 
     let callee_canister = opts.canister.as_str();
-    let canister_id_store = CanisterIdStore::for_env(env)?;
+    let canister_id_store =
+        CanisterIdStore::for_env(env).context("Failed to load canister id store.")?;
 
     let canister_id = Principal::from_text(callee_canister)
-        .or_else(|_| canister_id_store.get(callee_canister))?;
+        .or_else(|_| canister_id_store.get(callee_canister))
+        .with_context(|| format!("Failed to get canister id for {}.", callee_canister))?;
 
-    fetch_root_key_if_needed(env).await?;
+    fetch_root_key_if_needed(env)
+        .await
+        .context("Failed to fetch root key.")?;
     let controller_blob = agent
         .read_state_canister_info(canister_id, "controllers", false)
-        .await?;
+        .await
+        .with_context(|| format!("Failed to read canister {}'s controllers.", canister_id))?;
     let cbor: Value = serde_cbor::from_slice(&controller_blob)
         .map_err(|_| anyhow!("Invalid cbor data in controllers canister info."))?;
     let controllers = if let Value::Array(vec) = cbor {
         vec.into_iter()
             .map(|elem: Value| {
                 if let Value::Bytes(bytes) = elem {
-                    Ok(Principal::try_from(&bytes)?.to_text())
+                    Ok(Principal::try_from(&bytes)
+                        .context("Failed to construct principal.")?
```

Comment:
Even better would be to 
```suggestion
                        .context("Failed to construct principal of controller from bytes ({}).", hex::encode(&bytes))?
```